src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -610,21 +689,34 @@ private void CompletePipe()
             }
         }
 
-        internal bool IsReadAsyncCompleted => _readerAwaitable.IsCompleted;
+        internal ValueTaskSourceStatus GetReadAsyncStatus()
+        {
+            if (_readerAwaitable.IsCompleted)
+            {
+                if (_writerCompletion.IsFaulted)
+                {
+                    return ValueTaskSourceStatus.Faulted;
+                }
+                
+                return ValueTaskSourceStatus.Succeeded;
+            }
+            return ValueTaskSourceStatus.Pending;
+        }
 
-        internal void OnReadAsyncCompleted(Action continuation)
+        internal void OnReadAsyncCompleted(Action<object> continuation, object state, ValueTaskSourceOnCompletedFlags flags)
         {
-            Action awaitable;
+            Action<object> completion;
+            object completionState;
             bool doubleCompletion;
             lock (_sync)
             {
-                awaitable = _readerAwaitable.OnCompleted(continuation, out doubleCompletion);
+                _readerAwaitable.OnCompleted(continuation, state, flags, out completion, out completionState, out doubleCompletion);
             }
             if (doubleCompletion)
             {
                 Writer.Complete(ThrowHelper.CreateInvalidOperationException_NoConcurrentOperation());
             }
-            TrySchedule(_readerScheduler, awaitable);
+            TrySchedule(_readerScheduler, completion, completionState);
```

Comment:
I don't know if it's an issue with pipelines, but in general there are concerns about potential stack dives here. In the await pattern, if the operation completes between the IsCompleted check and the call to OnCompleted, we generally want OnCompleted to schedule asynchronously; otherwise if that repeatedly happened and every call ended up going deeper and deeper on the stack, we could overflow. Seems like this could be an issue if the inline scheduler is used.